db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/sqlgrammar.jj

```diff
@@ -10914,28 +10915,45 @@ tableReferenceTypes(boolean nestedInParens) throws StandardException :
 Object[]
 optionalTableClauses() throws StandardException :
 {
-    Object[]             otc = null;
-    Properties            tableProperties = null;
-    ResultColumnList    derivedRCL = null;
-    String                correlationName = null;
+    Object[]         otc = null;
+    Properties       tableProperties = null;
+    ResultColumnList derivedRCL = null;
+    String           correlationName = null;
+    Long             txnId = null;
 }
 {
     otc = optionalTableProperties()
     {
         otc[OPTIONAL_TABLE_CLAUSES_DERIVED_RCL] = derivedRCL;
         otc[OPTIONAL_TABLE_CLAUSES_CORRELATION_NAME] = correlationName;
+        otc[OPTIONAL_TABLE_CLAUSES_TXN_ID] = txnId;
         return otc;
     }
 |
+    LOOKAHEAD( { getToken(1).kind != AS || getToken(2).kind != OF })
     [ [ <AS> ]
```

Comment:
Could we remove the lookahead with something like

`[ [ <AS> [ <OF> txnId = asOfClause() [ AS ] ] ...`

Reply:
I tried but I couldn't find a good alternative, I mean with your proposal we would bind `AS OF` with having an alias (line 10936) which is not necessary.
Ideally we should allow the following cases:
* `<ALIAS>`
* `AS <ALIAS>`
* `<ALIAS> AS OF ...`
* `AS <ALIAS> AS OF ...`
* `AS OF ...`
I couldn't find single rule for all these cases that why I broke it in two.

Reply:
Yes @dgomezferro , you are right, the we need upgrade script to invalidate the saved plans in sys.systatements.